File: Ruby/test/test_SomeParsers.rb
require 'SomeParsers'

class TestSomeParsers < Test::Unit::TestCase

    def test_alnum
        assert_equal(alnum.parse("123abc  ").value, [['1', "23abc  "]])
        assert_equal(alnum.parse("_123  abc").value, [['_', "123  abc"]])
        assert_equal(alnum.parse("!@%$").value, [])
    end
    
    def test_char
        assert_equal(char('a').parse("abc").value, [['a', "bc"]])
        assert_equal(char('z').parse("abc").value, [])
    end

    def test_symbol
        assert_equal(symbol('+').parse(" + abc").value, [['+', "abc"]])
        assert_equal(symbol('+').parse("abc").value, [])
    end

    def test_name
        nsin = name_("sin")
        
        assert_equal(nsin.parse(" sin ").value, [["sin", ""]])
        assert_equal(nsin.parse("  sin  (1.)").value, [["sin", "(1.)"]])
        assert_equal(nsin.parse("sinabc").value, [])
    end

    def test_optional_s
        assert_equal(optional_s(name_("sin")).parse(" sin abc").value, [["sin", "abc"]])
        assert_equal(optional_s(name_("sin")).parse("abc").value, [["", "abc"]])
        assert_equal(optional_s(char('1')).parse("1234").value, [["1", "234"]])
        assert_equal(optional_s(char('1')).parse("abc").value, [["", "abc"]])
    end
    
    def test_sign
        assert_equal(sign.parse("abc").value, [["", "abc"]])
        assert_equal(sign.parse("+abc").value, [['+', "abc"]])
        assert_equal(sign.parse("-abc").value, [['-', "abc"]])
    end

    def test_digits
        assert_equal(digits.parse("123abc").value, [["123", "abc"]])
        assert_equal(digits.parse("123  abc").value, [["123", "  abc"]])
        assert_equal(digits.parse("abc").value, [["", "abc"]])
    end

    def test_double
        assert_equal(double.parse(" 1 abc").value, [[1, "abc"]])
        assert_equal(double.parse(" 1. abc").value, [[1, "abc"]])
        assert_equal(double.parse(" 1.23 abc").value, [[1.23, "abc"]])
        assert_equal(double.parse(" .23 abc").value, [[0.23, "abc"]])
        assert_equal(double.parse("1.23e10abc").value, [[1.23e10, "abc"]])
        assert_equal(double.parse("1.23e-10abc").value, [[1.23e-10, "abc"]])
        assert_equal(double.parse("abc").value, [])
    end

    def test_between
        expr = between(symbol('('), symbol(')')) { double }
        
        assert_equal(expr.parse(" ( 123 ) abc").value, [[123, "abc"]])
        assert_equal(expr.parse(" ( 123 abc").value, [])
        assert_equal(expr.parse(" 123 ) abc").value, [])
    end

    def test_chainlr1
        add = symbol('+').skip { Parser.pure(lambda { |x, y| x + y }) }
        sub = symbol('-').skip { Parser.pure(lambda { |x, y| x - y }) }
        pow = symbol('^').skip { Parser.pure(lambda { |x, y| Math.exp(y * Math.log(x)) }) }

        pexpr = chainl1(double, add | sub, false)

        assert_equal(pexpr.parse("7").value, [[7, ""]])
        assert_equal(pexpr.parse("7abc").value, [[7, "abc"]])
        assert_equal(pexpr.parse("7-1").value, [[6, ""]])
        assert_equal(pexpr.parse(" 7 - 1 - 2 abc").value, [[4, "abc"]])
        assert_equal(pexpr.parse(" 7 - 1 + 2 - 3 abc").value, [[5, "abc"]])
        assert_equal(pexpr.parse("abc").value, [])

        assert_equal(double.chainr1(pow).parse("3 ^ 2 ^ 3 abc")
            .transform { |pair| [pair[0].round, pair[1]] }.value, [[6561, "abc"]])
    end

end

File: Perl/test/test_some_parsers.pl
require "./src/some_parsers.pl";

use strict;
use warnings;
use Test::More; #tests => 2; # Declare the number of expected tests

sub test_char {
  is(SomeParsers->char('a')->parse("abc"), Maybe->Just(Pair->new('a', "bc")),
    "char('a')->parse(\"abc\") should be equal to Just(('a', \"bc\")");
  is(SomeParsers->char('z')->parse("abc"), Maybe->Nothing,
    "char('z')->parse(\"abc\") should be equal to Nothing");
  is(SomeParsers->char('a')->parse(""), Maybe->Nothing,
    "char('a')->parse(\"\") should be equal to Nothing");
}

sub test_symbol {
  is(SomeParsers->symbol('+')->parse(" + abc"), Maybe->Just(Pair->new('+', "abc")),
    "symbol('+')->parse(\" + abc\") should be equal to Just(('+', \"abc\"))");
  is(SomeParsers->symbol('+')->parse("abc"), Maybe->Nothing,
    "symbol('+')->parse(\"abc\") should be equal to Nothing");
}

sub test_alnum {
  is(SomeParsers->alnum->parse("123abc"), Maybe->Just(Pair->new('1', "23abc")),
    "alnum->parse(\"123abc\") should be equal to Just(('1', \"23abc\"))");
  is(SomeParsers->alnum->parse("abc"), Maybe->Just(Pair->new('a', "bc")),
    "alnum->parse(\"abc\") should be equal to Just(('a', \"bc\"))");
  is(SomeParsers->alnum->parse("_123abc"), Maybe->Just(Pair->new('_', "123abc")),
    "alnum->parse(\"_123abc\") should be equal to Just(('_', \"123abc\"))");
  is(SomeParsers->alnum->parse("!@#"), Maybe->Nothing,
    "alnum->parse(\"!@#\") should be equal to Nothing");
}

sub test_sign {
  is(SomeParsers->sign->parse("abc"), Maybe->Just(Pair->new("", "abc")),
    "sign->parse(\"abc\") should be equal to Just((\"\", \"abc\"))");
  is(SomeParsers->sign->parse("+abc"), Maybe->Just(Pair->new("+", "abc")),
    "sign->parse(\"+abc\") should be equal to Just((\"+\", \"abc\"))");
  is(SomeParsers->sign->parse("-abc"), Maybe->Just(Pair->new("-", "abc")),
    "sign->parse(\"-abc\") should be equal to Just((\"-\", \"abc\"))");

  is(SomeParsers->usign->parse("abc"), Maybe->Just(Pair->new("", "abc")),
    "usign->parse(\"abc\") should be equal to Just((\"\", \"abc\"))");
  is(SomeParsers->usign->parse(" + abc"), Maybe->Just(Pair->new("+", "abc")),
    "usign->parse(\" + abc\") should be equal to Just((\"+\", \"abc\"))");
  is(SomeParsers->usign->parse(" - abc"), Maybe->Just(Pair->new("-", "abc")),
    "usign->parse(\" - abc\") should be equal to Just((\"-\", \"abc\"))");
}

sub test_digits {
  is(SomeParsers->digits->parse("123abc"), Maybe->Just(Pair->new("123", "abc")),
    "digits->parse(\"123abc\") should be equal to Just((\"123\", \"abc\"))");
  is(SomeParsers->digits->parse("123  abc"), Maybe->Just(Pair->new("123", "  abc")),
    "digits->parse(\"123  abc\") should be equal to Just((\"123\", \"  abc\"))");
  is(SomeParsers->digits->parse("abc"), Maybe->Just(Pair->new("", "abc")),
    "digits->parse(\"abc\") should be equal to Just((\"\", \"abc\"))");
}

sub test_double {
  is(SomeParsers->double->parse("1 abc"), Maybe->Just(Pair->new(1.0, "abc")),
    "double->parse(\"1 abc\") should be equal to Just((1.0, \"abc\"))");
  is(SomeParsers->double->parse("1. abc"), Maybe->Just(Pair->new(1.0, "abc")),
    "double->parse(\"1. abc\") should be equal to Just((1.0, \"abc\"))");
  is(SomeParsers->double->parse("1.23 abc"), Maybe->Just(Pair->new(1.23, "abc")),
    "double->parse(\"1.23 abc\") should be equal to Just((1.23, \"abc\"))");
  is(SomeParsers->double->parse("-1.23 abc"), Maybe->Nothing,
    "double->parse(\"-1.23 abc\") should be equal to Nothing");
  is(SomeParsers->double->parse(".23 abc"), Maybe->Just(Pair->new(0.23, "abc")),
    "double->parse(\".23 abc\") should be equal to Just((0.23, \"abc\"))");
  is(SomeParsers->double->parse(" + 1.23 abc"), Maybe->Nothing,
    "double->parse(\" + 1.23 abc\") should be equal to Nothing");
  is(SomeParsers->double->parse("1.23e10abc"), Maybe->Just(Pair->new(1.23e10, "abc")),
    "double->parse(\"1.23e10abc\") should be equal to Just((1.23e10, \"abc\"))");
  is(SomeParsers->double->parse("1.23e-10abc"), Maybe->Just(Pair->new(1.23e-10, "abc")),
    "double->parse(\"1.23e-10abc\") should be equal to Just((1.23e-10, \"abc\"))");
  is(SomeParsers->double->parse("abc"), Maybe->Nothing,
    "double->parse(\"abc\") should be equal to Nothing")
}

sub test_between {
  my $expr = Parser->between(SomeParsers->symbol('('), SomeParsers->symbol(')'), sub { SomeParsers->double });

  is($expr->parse(" ( 123 ) abc"), Maybe->Just(Pair->new(123.0, "abc")),
    "expr->parse(\" ( 123 ) abc\") should be equal to Just((123.0, \"abc\"))");
  is($expr->parse(" ( 123 abc"), Maybe->Nothing,
    "expr->parse(\" ( 123 abc\") should be equal to Nothing");
  is($expr->parse(" 123 ) abc"), Maybe->Nothing,
    "expr->parse(\" 123 ) abc\") should be equal to Nothing");
}

sub test_chainlr1 {
  my $add = SomeParsers->symbol('+') >> sub { Parser->pure(sub { shift() + shift }) };
  my $sub = SomeParsers->symbol('-') >> sub { Parser->pure(sub { shift() - shift }) };
  my $pow = SomeParsers->symbol('^') >> sub { Parser->pure(sub { exp(log(shift) * shift) }) };

  my $pexpr = SomeParsers->double->chainl1($add | sub { $sub }, 0);

  is($pexpr->parse("7abc"), Maybe->Just(Pair->new(7.0, "abc")),
    "expr->parse(\"7abc\") should be equal to Just((7.0, \"abc\"))");
  is($pexpr->parse(" 7 - 1 - 2 abc"), Maybe->Just(Pair->new(4.0, "abc")),
    "expr->parse(\" 7 - 1 - 2 abc\") should be equal to Just((4.0, \"abc\"))");
  is($pexpr->parse(" 7 - 1 + 2 - 3 abc"), Maybe->Just(Pair->new(5.0, "abc")),
    "expr->parse(\" 7 - 1 + 2 - 3 abc\") should be equal to Just((5.0, \"abc\"))");
  is($pexpr->parse("abc"), Maybe->Nothing, "expr->parse(\"abc\") should be equal to Nothing");
  is(SomeParsers->double->chainr1($pow)->parse("3 ^ 2 ^ 3 abc")->fmap(sub {
      my $pair = shift; Pair->new(int($pair->{first} + 0.5), $pair->{second});
    }), Maybe->Just(Pair->new(6561.0, "abc")),
    "expr->parse(\"3 ^ 2 ^ 3 abc\") should be equal to Just((6561.0, \"abc\"))");
}

test_char;
test_symbol;
test_alnum;
test_sign;
test_digits;
test_double;
test_between;
test_chainlr1;
done_testing;
